### ISSUE TITLE
[3.3.5][Warlock] Conflagrate - wrong damage

### ISSUE THREAD

**Issue by Keader** [TrinityCore/TrinityCore#14918]
Talent Description: Consumes an Immolate or Shadowflame effect on the enemy target to instantly deal damage equal to 60% of your Immolate or Shadowflame, and causes an additional 40% damage over 6 sec.

http://wotlk.openwow.com/spell=17962

Problem: 

With hits from 213 in imolate: 5x213 = 1065.
60% of 1065 should be 639, but the value obtained is 472.
Using the addon DrDamage, the damage should be 641 (according to the addon)

Build Used:
![build lock](https://cloud.githubusercontent.com/assets/7120913/8408069/5fff8532-1e43-11e5-9fef-6a941be3e902.jpg)

Tests conducted without spell power and talent above (necessary talents to release the conflagrate)
It may have something to do with #14869 (https://github.com/TrinityCore/TrinityCore/issues/14869#issuecomment-112065926)

Test on rev 170ba68e63c46323f8624ecff24e6de1d38bd97b

**Comment by Keader** [comment]
@robinsch When you have time, could you take a look?

**Comment by Keader** [comment]
@robinsch When you have time, could you take a look?

**Comment by Keader** [comment]
Apparently the problem is with the value of the variable "pdamage". Printing the value of the variable is leaving to less than 213 (which would be the damage each tick of immolate)
@robinsch You could take a look?

**Comment by mik1893** [comment]
just try as he did for the other spell...
```diff
diff --git a/src/server/game/Spells/SpellEffects.cpp b/src/server/game/Spells/SpellEffects.cpp
index 965b246..2c31159 100644
--- a/src/server/game/Spells/SpellEffects.cpp
+++ b/src/server/game/Spells/SpellEffects.cpp
@@ -422,6 +422,7 @@ void Spell::EffectSchoolDMG(SpellEffIndex effIndex)
                     {
                         uint32 pdamage = uint32(std::max(aura->GetAmount(), 0));
                         pdamage = m_caster->SpellDamageBonusDone(unitTarget, aura->GetSpellInfo(), pdamage, DOT, aura->GetBase()->GetStackAmount());
+                        pdamage *= m_caster->SpellDamagePctDone(unitTarget, aura->GetSpellInfo(), SPELL_DIRECT_DAMAGE);
                         pdamage = unitTarget->SpellDamageBonusTaken(m_caster, aura->GetSpellInfo(), pdamage, DOT, aura->GetBase()->GetStackAmount());
                         uint32 pct_dir = m_caster->CalculateSpellDamage(unitTarget, m_spellInfo, (effIndex + 1));
                         uint8 baseTotalTicks = uint8(m_caster->CalcSpellDuration(aura->GetSpellInfo()) / aura->GetSpellInfo()->Effects[EFFECT_0].Amplitude);
```

**Comment by Nyeriah** [comment]
https://github.com/TrinityCore/TrinityCore/pull/14898

**Comment by Keader** [comment]
@mik1893 I had already tried to do this yesterday ... for this spell does not work

**Comment by robinsch** [comment]
I will check it later.

**Comment by mik1893** [comment]
just to make it an all in one if you want to push a PR, @robinsch :+1: 
- swiftmend ( you already posted working code)
- chimera shot
- conflagrate

**Comment by Keader** [comment]
I tried to find other spells that have similar operation, but can not think of any other xD

**Comment by robinsch** [comment]
@Keader: Can you try to revert the changes in Unit.cpp https://github.com/TrinityCore/TrinityCore/commit/198ffba5ee3b64852b3b090ebdeebd87494cc9f6 and test it again? I think this is caused by changes in this PR, if so @Shauren do you think that it's better to keep changes in Unit.cpp or fixup those few spells by adding ApplySpellMod to them.

**Comment by ccrs** [comment]
ummm, i think those calculation errors where before ur general fix on unit.cpp (no evidences though)

**Comment by Keader** [comment]
@robinsch  Giving revert in your fix appears the following problem:
Damage from the DoT Imolate, falls 223.5 (Damage Correct according to DRdamage) for 157.
The damage conflagrate rises of 472 (damage wrong) to 640 (very near of what should be the correct one)
In short ... fix the direct damage, but affect the damage of the DoT

**Comment by robinsch** [comment]
@Keader: Did you revert the full commit or just changes in Unit.cpp?

```diff
diff --git a/src/server/game/Spells/SpellEffects.cpp b/src/server/game/Spells/SpellEffects.cpp
index 3829e12..e81bcd4 100644
--- a/src/server/game/Spells/SpellEffects.cpp
+++ b/src/server/game/Spells/SpellEffects.cpp
@@ -423,6 +423,8 @@ void Spell::EffectSchoolDMG(SpellEffIndex effIndex)
                         uint32 pdamage = uint32(std::max(aura->GetAmount(), 0));
                         pdamage = m_caster->SpellDamageBonusDone(unitTarget, aura->GetSpellInfo(), pdamage, DOT, aura->GetBase()->GetStackAmount());
                         pdamage = unitTarget->SpellDamageBonusTaken(m_caster, aura->GetSpellInfo(), pdamage, DOT, aura->GetBase()->GetStackAmount());
+                        if (Player* modOwner = m_caster->GetSpellModOwner())
+                            modOwner->ApplySpellMod(GetSpellInfo()->Id, SPELLMOD_DOT, pdamage);
                         uint32 pct_dir = m_caster->CalculateSpellDamage(unitTarget, m_spellInfo, (effIndex + 1));
                         uint8 baseTotalTicks = uint8(m_caster->CalcSpellDuration(aura->GetSpellInfo()) / aura->GetSpellInfo()->Effects[EFFECT_0].Amplitude);
                         damage += int32(CalculatePct(pdamage * baseTotalTicks, pct_dir));
@@ -1413,7 +1415,11 @@ void Spell::EffectHeal(SpellEffIndex /*effIndex*/)
 
             int32 tickheal = targetAura->GetAmount();
             if (Unit* auraCaster = targetAura->GetCaster())
+            {
                 tickheal = auraCaster->SpellHealingBonusDone(unitTarget, targetAura->GetSpellInfo(), tickheal, DOT);
+                if (Player* modOwner = auraCaster->GetSpellModOwner())
+                    modOwner->ApplySpellMod(GetSpellInfo()->Id, SPELLMOD_DOT, tickheal);
+            }
             //int32 tickheal = targetAura->GetSpellInfo()->EffectBasePoints[idx] + 1;
             //It is said that talent bonus should not be included
 diff --git a/src/server/scripts/Spells/spell_hunter.cpp b/src/server/scripts/Spells/spell_hunter.cpp
index 75df264..d59a58a 100644
--- a/src/server/scripts/Spells/spell_hunter.cpp
+++ b/src/server/scripts/Spells/spell_hunter.cpp
@@ -213,6 +213,8 @@ class spell_hun_chimera_shot : public SpellScriptLoader
                                 basePoint = (aurEff->GetAmount() + aurEff->GetBonusAmount()) * aurEff->GetDonePct();
                                 ApplyPct(basePoint, TickCount * 40);
                                 basePoint = unitTarget->SpellDamageBonusTaken(caster, aura->GetSpellInfo(), basePoint, DOT, aura->GetStackAmount());
+                                if (Player* modOwner = caster->GetSpellModOwner())
+                                    modOwner->ApplySpellMod(GetSpellInfo()->Id, SPELLMOD_DOT, basePoint);
 
                                 aurEff->SetBonusAmount(caster->SpellDamageBonusDone(unitTarget, aurEff->GetSpellInfo(), 0, DOT));
                             }
```

**Comment by Keader** [comment]
@robinsch He had done revert the entire commit.

On the new fix did not work in any of the three spells.
The old version of swiftmend ( https://github.com/TrinityCore/TrinityCore/issues/14958#issuecomment-116285729 ) works, however the new no.
In 3 cases the damage of spells there was no change

**Comment by robinsch** [comment]
@Keader: I told you to try it with just reverting the changes done to Unit.cpp in https://github.com/TrinityCore/TrinityCore/commit/198ffba5ee3b64852b3b090ebdeebd87494cc9f6.

**Comment by Keader** [comment]
@robinsch :
My mistake, sorry.

Test without any fix, applying only revert the Unit.cpp (198ffba5ee3b64852b3b090ebdeebd87494cc9f6)

Conflagrate: DoT Damage to the Imolate (OK) Direct Damage (OK)
Swiftmend: Healing Over Time (OK) Direct Heal (OK)
Chimera Shot: DoT Damage of the Serpent Sting (OK) Direct Damage (WRONG)

2 Spells returned to work perfectly, however the Chimera Shot is still wrong with direct damage.

**Comment by ccrs** [comment]
Corruption damage is wrong without the Unit.cpp fix, isnt it?

**Comment by robinsch** [comment]
Let's move the conversation to https://github.com/TrinityCore/TrinityCore/pull/14985

**Comment by ccrs** [comment]
I guess we return the conversation here xd

If Unit.cpp changes are reverted, what about this https://github.com/TrinityCore/TrinityCore/issues/14817?

**Comment by Keader** [comment]
I'm compiling to test here

**Comment by robinsch** [comment]
It will still work.

**Comment by mik1893** [comment]
so is corruption still working properly with the revert? :P

**Comment by Keader** [comment]
Yes, reversing only the unit.cpp

**Comment by mik1893** [comment]
Closed by #14987
